### PR TITLE
[portal_sale] Migrate into new API

### PR DESCRIPTION
**Pull request by gan-odoo**
Task : https://www.odoo.com/web#id=20767&view_type=form&model=project.task&menu_id=3942&action=327
Pad : https://pad.odoo.com/p/r.pSixt0tTeBt9sVtH

### CODE REVIEW

File: addons/portal_sale/models/sale_order.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        '''  Override to use a modified template that includes a portal signup link '''
+        action_dict = super(SaleOrder, self).action_quotation_send()
+        try:
+            template = self.env.ref('portal_sale.email_template_edi_sale')
+            # assume context is still a dict, as prepared by super
+            ctx = action_dict['context']
+            ctx['default_template_id'] = template.id
+            ctx['default_use_template'] = True
+        except Exception:
+            pass
```
Comment (xmo-odoo) [inline addons/portal_sale/models/sale_order.py]: What's that "oh let's just ignore any and all exception" doing here?

File: addons/portal_sale/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        # fetch the partner's id and subscribe the partner to the sale order
+        partner = self.partner_id
+        if partner not in self.message_partner_ids:
+            self.message_subscribe([partner.id])
+        return super(SaleOrder, self).action_confirm()
+
+    @api.multi
+    def get_signup_url(self):
+        contex_signup = dict(self.env.context, signup_valid=True)
+        return self.partner_id.with_context(contex_signup)._get_signup_url_for_action(
```
Comment (xmo-odoo) [inline addons/portal_sale/models/sale_order.py]: `self.partner_id.with_context(signup_valid=True)`

File: addons/portal_sale/models/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            action='mail.action_mail_redirect',
+            model=self._name, res_id=self.id)[self.partner_id.id]
```
Comment (xmo-odoo) [inline addons/portal_sale/models/sale_order.py]: If parameters use an indented style, put them all on individual lines

File: addons/portal_sale/models/sale_order.py (hunk 1 above)
Comment (mart-e) [inline addons/portal_sale/models/sale_order.py]: `self.env.ref('portal_sale.email_template_edi_sale', raise_if_not_found=False)`

File: addons/portal_sale/models/account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from openerp import api, models
+
+class AccountInvoice(models.Model):
```
Comment (xmo-odoo) [inline addons/portal_sale/models/account_invoice.py]: See commends below for other file with almost identical methods.

File: addons/portal_sale/views/account_invoice_views.xml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    </record>
+
+    <record id="portal_action_invoices_form_spec" model="ir.actions.act_window.view">
+        <field name="act_window_id" ref="portal_action_invoices"/>
+        <field name="view_id" ref="account.invoice_form"/>
+        <field name="view_mode">form</field>
+        <field name="sequence" eval="1"/>
+    </record>
+
+    <menuitem name="Invoices" id="portal_invoices" parent="portal.portal_orders"
```
Comment (WHPODOO) [inline addons/portal_sale/views/account_invoice_views.xml]: Id should be first : 
<menuitem id="portal_invoices" name="Invoices"  parent="portal.portal_orders" ...

File: addons/portal_sale/models/account_invoice.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    _inherit = 'account.invoice'
+
+    @api.multi
+    def action_invoice_sent(self):
+        '''  Override to use a modified template that includes a portal signup link '''
+        action_dict = super(AccountInvoice, self).action_invoice_sent()
+        template = self.env.ref('portal_sale.email_template_edi_invoice', raise_if_not_found=False)
+        # assume context is still a dict, as prepared by super
+        ctx = action_dict['context']
+        ctx['default_template_id'] = template and template.id or None
```
Comment (mdi-odoo) [inline addons/portal_sale/models/account_invoice.py]: Here template.id is enough .

File: addons/portal_sale/models/account_invoice.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        ctx['default_use_template'] = True
+        return action_dict
+
+    @api.multi
+    def invoice_validate(self):
+        # fetch the partner's id and subscribe the partner to the invoice
+        partner = self.partner_id
```
Comment (mdi-odoo) [inline addons/portal_sale/models/account_invoice.py]: Use self.ensure_one() to verify that the self contains only one record.

File: addons/portal_sale/models/account_invoice.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        if partner not in self.message_partner_ids:
+            self.message_subscribe([partner.id])
+        return super(AccountInvoice, self).invoice_validate()
+
+    @api.multi
+    def get_signup_url(self):
+        return self.partner_id.with_context(signup_valid=True)._get_signup_url_for_action(
```
Comment (mdi-odoo) [inline addons/portal_sale/models/account_invoice.py]: Use self.ensure_one() to verify that the self contains only one record.

File: addons/portal_sale/models/account_invoice.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            action='mail.action_mail_redirect',
+            model=self._name,
+            res_id=self.id)[self.partner_id.id]
+
+    @api.multi
+    def get_formview_action(self):
+        if self.env.user.share:
```
Comment (mdi-odoo) [inline addons/portal_sale/models/account_invoice.py]: Same as above.

File: addons/portal_sale/models/mail.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+from openerp import api, models
+
+class Mail(models.Model):
+    _inherit = 'mail.mail'
+
+    @api.multi
+    def _postprocess_sent_message_v9(self, mail_sent=True):
+        if mail_sent and self.model == 'sale.order':
+            sale_order = self.env['sale.order'].browse(self.res_id)
```
Comment (mdi-odoo) [inline addons/portal_sale/models/mail.py]: Why you need to rebrowse, Here 'self.res_id' is already a browse record.

File: addons/portal_sale/models/sale_order.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    _inherit = 'sale.order'
+
+    @api.multi
+    def action_quotation_send(self):
+        '''  Override to use a modified template that includes a portal signup link '''
+        action_dict = super(SaleOrder, self).action_quotation_send()
+        template = self.env.ref('portal_sale.email_template_edi_sale', raise_if_not_found=False)
+        # assume context is still a dict, as prepared by super
+        ctx = action_dict['context']
+        ctx['default_template_id'] = template and template.id or None
```
Comment (mdi-odoo) [inline addons/portal_sale/models/sale_order.py]: Here template.id is enough .

File: addons/portal_sale/models/sale_order.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        ctx['default_use_template'] = True
+        return action_dict
+
+    @api.multi
+    def action_confirm(self):
+        # fetch the partner's id and subscribe the partner to the sale order
+        partner = self.partner_id
```
Comment (mdi-odoo) [inline addons/portal_sale/models/sale_order.py]: Use self.ensure_one() to verify that the self contains only one record .

File: addons/portal_sale/models/sale_order.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        if partner not in self.message_partner_ids:
+            self.message_subscribe([partner.id])
+        return super(SaleOrder, self).action_confirm()
+
+    @api.multi
+    def get_signup_url(self):
+        return self.partner_id.with_context(signup_valid=True)._get_signup_url_for_action(
```
Comment (mdi-odoo) [inline addons/portal_sale/models/sale_order.py]: Same as above .

File: addons/portal_sale/models/sale_order.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+            action='mail.action_mail_redirect',
+            model=self._name,
+            res_id=self.id)[self.partner_id.id]
+
+    @api.multi
+    def get_formview_action(self):
+        if self.env.user.share:
```
Comment (mdi-odoo) [inline addons/portal_sale/models/sale_order.py]: Same as above .

File: addons/portal_sale/models/mail.py (hunk 10 above)
Comment (gan-odoo) [inline addons/portal_sale/models/mail.py]: @mdi-odoo here 'self.res_id' is integer id not a browse record.

File: addons/portal_sale/models/account_invoice.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        # assume context is still a dict, as prepared by super
+        ctx = action_dict['context']
+        ctx['default_template_id'] = template.id or None
+        ctx['default_use_template'] = True
+        return action_dict
+
+    @api.multi
+    def invoice_validate(self):
+        # fetch the partner's id and subscribe the partner to the invoice
+        self.ensure_one()
```
Comment (mdi-odoo) [inline addons/portal_sale/models/account_invoice.py]: Make the method to work on multiple record(s).

File: addons/portal_sale/models/sale_order.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+    @api.multi
+    def action_quotation_send(self):
+        '''  Override to use a modified template that includes a portal signup link '''
+        self.ensure_one()
+        action_dict = super(SaleOrder, self).action_quotation_send()
+        template = self.env.ref('portal_sale.email_template_edi_sale', raise_if_not_found=False)
+        # assume context is still a dict, as prepared by super
+        ctx = action_dict['context']
+        ctx['default_template_id'] = template.id or None
```
Comment (mdi-odoo) [inline addons/portal_sale/models/sale_order.py]: Just 'template.id'  is enough .

File: addons/portal_sale/models/sale_order.py [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+        ctx['default_use_template'] = True
+        return action_dict
+
+    @api.multi
+    def action_confirm(self):
+        # fetch the partner's id and subscribe the partner to the sale order
+        self.ensure_one()
```
Comment (mdi-odoo) [inline addons/portal_sale/models/sale_order.py]: Make the method to work on multiple record(s).

File: addons/portal_sale/models/mail.py [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+            partner = sale_order.partner_id
+            # Add the customer in the SO as follower
+            if partner not in sale_order.message_partner_ids:
+                sale_order.message_subscribe([partner.id])
+            # Add all recipients of the email as followers
+            for partners in self.partner_ids:
+                if partners in sale_order.message_partner_ids:
+                    sale_order.message_subscribe([partners.id])
```
Comment (mdi-odoo) [inline addons/portal_sale/models/mail.py]: Use if partners **not** in sale_order.message_partner_ids .

File: addons/portal_sale/models/mail.py [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+    def _postprocess_sent_message_v9(self, mail_sent=True):
+        self.ensure_one()
+        if mail_sent and self.model == 'sale.order':
+            sale_order = self.env['sale.order'].browse(self.res_id)
+            partner = sale_order.partner_id
+            # Add the customer in the SO as follower
+            if partner not in sale_order.message_partner_ids:
+                sale_order.message_subscribe([partner.id])
+            # Add all recipients of the email as followers
+            for partners in self.partner_ids:
```
Comment (mdi-odoo) [inline addons/portal_sale/models/mail.py]: Only 'partner' not 'partners' .

File: addons/portal_sale/views/account_config_settings_views.xml [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+<odoo>
+
+        <!--  Add payment options to sale.order and invoice forms -->
+        <record id="portal_sale_payment_option_config" model="ir.ui.view">
+                <field name="model">account.config.settings</field>
+                <field name="inherit_id" ref="account.view_account_config_settings"/>
+                <field name="arch" type="xml">
+                    <xpath expr="//group[@name='bank_cash']/div" position="inside">
+                        <div>
+                            <field name="group_payment_options" class="oe_inline"/>
```
Comment (hmo-odoo) [inline addons/portal_sale/views/account_config_settings_views.xml]: use o_row in div instead of oe_inline

**Pull request merged**